cli/src/program.rs

```diff
@@ -1103,6 +1331,162 @@ fn process_dump(
+        &[bpf_loader_upgradeable::close(
+            &account_pubkey,
+            &recipient_pubkey,
+            &authority_signer.pubkey(),
+        )],
+        Some(&config.signers[0].pubkey()),
+    ));
+
+    tx.try_sign(&[config.signers[0], authority_signer], blockhash)?;
+    let result = rpc_client.send_and_confirm_transaction_with_spinner_and_config(
+        &tx,
+        config.commitment,
+        RpcSendTransactionConfig {
+            skip_preflight: true,
+            preflight_commitment: Some(config.commitment.commitment),
+            ..RpcSendTransactionConfig::default()
+        },
+    );
+    if let Err(err) = result {
+        if let ClientErrorKind::TransactionError(TransactionError::InstructionError(
+            _,
+            InstructionError::InvalidInstructionData,
+        )) = err.kind()
+        {
+            return Err("Closing a buffer account is not supported by the cluster".into());
+        } else {
+            return Err(format!("Close failed: {}", err).into());
+        }
+    }
+    Ok(())
+}
+
+fn process_close(
+    rpc_client: &RpcClient,
+    config: &CliConfig,
+    account_pubkey: Option<Pubkey>,
+    recipient_pubkey: Pubkey,
+    authority_index: SignerIndex,
+    use_lamports_unit: bool,
+) -> ProcessResult {
+    let authority_signer = config.signers[authority_index];
+    let mut buffers = vec![];
+
+    if let Some(account_pubkey) = account_pubkey {
+        if let Some(account) = rpc_client
+            .get_account_with_commitment(&account_pubkey, config.commitment)?
+            .value
+        {
+            if let Ok(UpgradeableLoaderState::Buffer { authority_address }) = account.state() {
+                if authority_address != Some(authority_signer.pubkey()) {
+                    return Err(format!(
+                        "Buffer account authority {:?} does not match {:?}",
+                        authority_address,
+                        Some(authority_signer.pubkey())
+                    )
+                    .into());
+                } else {
+                    close(
+                        rpc_client,
+                        config,
+                        &account_pubkey,
+                        &recipient_pubkey,
+                        authority_signer,
+                    )?;
+
+                    if let Ok(UpgradeableLoaderState::Buffer { authority_address }) =
+                        account.state()
+                    {
+                        buffers.push(CliUpgradeableBuffer {
+                            address: account_pubkey.to_string(),
+                            authority: authority_address
+                                .map(|pubkey| pubkey.to_string())
+                                .unwrap_or_else(|| "none".to_string()),
+                            data_len: 0,
+                            lamports: account.lamports,
+                            use_lamports_unit,
+                        });
```

Comment:
I'm not sure I understand what this is for. Is the intent to show that the buffer was closed?

Reply:
Correct, to show the caller which accounts were closed

Reply:
Ok, I was confused why you deserialize the account state a second time without refetching the account in between.

Reply:
Ha, blind spot